File: ios/LivingRoomEdge/LivingRoomEdge/Edge/SkillRegistry.swift
import Foundation

final class SkillRegistry {
    private var skills: [String: Skill] = [:]

    func register(_ skill: Skill) {
        skills[skill.service().serviceId] = skill
    }

    func registerAll(_ items: [Skill]) {
        items.forEach { register($0) }
    }

    /// Replace entire skill map (used when rebuilding from Capability plugins).
    func replaceAll(_ items: [Skill]) {
        skills.removeAll(keepingCapacity: true)
        registerAll(items)
    }

    func get(_ skillId: String) -> Skill? {
        skills[skillId]
    }

    func findByCapability(_ capabilityId: String) -> (Skill, ServiceDescriptor)? {
        let want = capabilityId.trimmingCharacters(in: .whitespacesAndNewlines)
        guard !want.isEmpty else { return nil }
        for skill in skills.values {
            let svc = skill.service()
            if svc.capabilities.contains(where: { $0.capabilityId == want }) {
                return (skill, svc)
            }
        }
        return nil
    }

    func all() -> [Skill] {
        Array(skills.values)
    }

    func services() -> [ServiceDescriptor] {
        skills.values.map { $0.service() }.sorted { $0.serviceId < $1.serviceId }
    }
}

File: android/living-room-android/src/main/java/com/smarthome/livingroom_android/edge/SkillRegistry.kt
package com.smarthome.livingroom_android.edge

import com.smarthome.livingroom_android.brain.dto.ServiceDescriptor
import com.smarthome.livingroom_android.skill.Skill

class SkillRegistry {
    private val skills = linkedMapOf<String, Skill>()

    fun register(skill: Skill) {
        val id = skill.service().serviceId
        skills[id] = skill
    }

    fun registerAll(vararg skills: Skill) {
        skills.forEach { register(it) }
    }

    fun get(serviceId: String): Skill? = skills[serviceId]

    /** Resolve skill that owns [capabilityId]. */
    fun findByCapability(capabilityId: String): Pair<Skill, ServiceDescriptor>? {
        val want = capabilityId.trim()
        if (want.isEmpty()) return null
        for (skill in skills.values) {
            val svc = skill.service()
            if (svc.capabilities.any { it.capabilityId == want }) {
                return skill to svc
            }
        }
        return null
    }

    fun services(): List<ServiceDescriptor> = skills.values.map { it.service() }

    fun all(): Collection<Skill> = skills.values
}
